### PR TITLE
Fix partition predicate numeric comparison (IT-036)

### LINKED ISSUE #85: Issue: Partition Predicate WHERE Clause - Unsupported and Broken Operators
# Issue: Partition Predicate WHERE Clause - Unsupported and Broken Operators

## Summary

The custom SQL commands that accept WHERE clauses for partition filtering (DROP PARTITIONS, MERGE SPLITS, PREWARM CACHE) have issues with certain SQL operators. Some operators don't work correctly due to type coercion issues, and others are not supported at all.

## Affected Commands

1. `DROP INDEXTABLES PARTITIONS FROM <table> WHERE <predicates>`
2. `MERGE SPLITS <table> WHERE <predicates> ...`
3. `PREWARM INDEXTABLES CACHE <table> WHERE <predicates> ...`

### 1. BETWEEN Operator - Broken for Numeric Comparisons

**Symptom**: `WHERE month BETWEEN 1 AND 6` produces incorrect results for partition values >= 10.

**Root Cause**: All partition values are stored as strings in the transaction log. The `resolveExpression` function in `PartitionPredicateUtils.scala:144-159` converts numeric literals to strings, but comparison then uses **lexicographic ordering** instead of numeric ordering.

**Example**:
```sql
-- Table has partitions: month = 1, 2, 3, ..., 10, 11, 12
DROP INDEXTABLES PARTITIONS FROM '/path' WHERE month BETWEEN 2 AND 11

-- Expected: Drop partitions 2, 3, 4, 5, 6, 7, 8, 9, 10, 11
-- Actual: Drops only 2, 3, 4, 5, 6, 7, 8, 9 (and maybe 10, 11)
-- Why: String comparison "10" < "2" because '1' < '2' lexicographically
```

**Code Location**: `PartitionPredicateUtils.scala:144-159`

```scala
def resolveExpression(expression: Expression, schema: StructType): Expression =
  expression.transform {
    case literal: org.apache.spark.sql.catalyst.expressions.Literal =>
      literal.dataType match {
        case StringType => literal
        case _ =>
          // BUG: Converts to string, but comparisons become lexicographic
          org.apache.spark.sql.catalyst.expressions.Literal(
            UTF8String.fromString(literal.value.toString), StringType)
      }
  }
```

### 2. Range Comparisons (>, <, >=, <=) - Same Issue

Any numeric comparison suffers from the same lexicographic ordering problem:

```sql
-- Broken examples:
WHERE month > 5    -- "10" is NOT > "5" lexicographically
WHERE year < 2023  -- "999" IS < "2023" lexicographically (wrong!)
WHERE day >= 10    -- "9" IS >= "10" lexicographically (wrong!)
```

### 3. IN Operator - Partially Working

The IN operator may work for string values but has edge cases:

```sql
-- Works (string comparisons):
WHERE region IN ('us-east', 'us-west')

-- May have issues (type coercion):
WHERE month IN (1, 2, 3)  -- Depends on how literals are handled
```

### 4. NOT Operator - Not Tested

The NOT operator may work at the grammar level but hasn't been validated:

```sql
WHERE NOT (region = 'us-east')
WHERE region NOT IN ('us-east', 'us-west')
```

### 5. Unsupported Operators

The following operators are **NOT supported** and either fail parsing or produce unexpected results:

| Operator | Example | Status |
|----------|---------|--------|
| `LIKE` | `WHERE region LIKE 'us-%'` | Not tested, may fail |
| `RLIKE` | `WHERE region RLIKE '^us-.*'` | Not tested, may fail |
| `<>` / `!=` | `WHERE region <> 'us-east'` | Not tested |
| Subqueries | `WHERE region IN (SELECT ...)` | Not supported |
| Functions | `WHERE UPPER(region) = 'US-EAST'` | Not supported |

### 6. Grammar Ambiguity with `.*?` Predicate Token

The grammar uses a lazy wildcard for predicate parsing:

```antlr
// IndexTables4SparkSqlBase.g4:92-94
predicateToken
    : .*?
    ;
```

This can cause issues when the WHERE clause precedes other optional clauses:

```sql
-- Potential parsing issue:
MERGE SPLITS '/path' WHERE date = '2024' AND status = 'active' TARGET SIZE 100M

-- The parser might incorrectly include "TARGET SIZE 100M" in the WHERE clause
-- or truncate the WHERE clause prematurely
```

### Solution 1: Numeric-Aware Comparison (Recommended)

Enhance `resolveExpression` to detect numeric partition columns and use appropriate comparison:

```scala
def resolveExpression(expression: Expression, schema: StructType): Expression = {
  expression.transform {
    case cmp @ GreaterThan(left, right) =>
      if (isNumericComparison(left, right, schema)) {
        GreaterThan(toNumeric(left), toNumeric(right))
      } else cmp
    case cmp @ LessThan(left, right) =>
      // Similar handling
    // ... handle all comparison operators
  }
}

private def isNumericComparison(left: Expression, right: Expression, schema: StructType): Boolean = {
  // Check if the partition column appears to contain numeric values
  // Based on actual partition values in the table
}
```

### Solution 2: Explicit Type Casting in Predicates

Require users to cast partition values explicitly when using numeric comparisons:

```sql
-- Instead of:
WHERE month BETWEEN 1 AND 6

-- Require:
WHERE CAST(month AS INT) BETWEEN 1 AND 6
```

This requires parsing support for CAST expressions.

### Solution 3: Numeric Partition Column Metadata

Store partition column types in table metadata and use them for proper comparison:

```scala
val partitionSchema = metadata.partitionSchema  // StructType with actual types
// Use the actual types instead of assuming StringType
```

### Solution 4: Fix Grammar for WHERE Clause

Replace the lazy wildcard with a proper expression grammar:

```antlr
// Instead of:
predicateToken: .*? ;

// Use:
predicateToken: booleanExpression ;

booleanExpression
    : booleanExpression AND booleanExpression
    | booleanExpression OR booleanExpression
    | NOT booleanExpression
    | '(' booleanExpression ')'
    | comparison
    | identifier BETWEEN value AND value
    | identifier IN '(' valueList ')'
    ;
```

## Test Cases to Add

```scala
test("BETWEEN with two-digit numeric partition values") {
  // Create table with months 1-12
  // DROP WHERE month BETWEEN 2 AND 11
  // Verify months 2-11 are dropped, NOT just 2-9
}

test("Greater than with numeric partition values") {
  // CREATE with values 1, 5, 10, 15, 20
  // DROP WHERE value > 5
  // Verify 10, 15, 20 are dropped (not just 6-9)
}

test("Less than with string year values") {
  // CREATE with years 2020, 2021, 2022, 2023
  // DROP WHERE year < '2022'
  // Verify only 2020, 2021 are dropped
}

test("IN operator with mixed values") {
  // Verify IN works correctly with numeric-looking partition values
}

test("NOT operator") {
  // Verify NOT negation works correctly
}
```

## Files to Modify

1. **`src/main/scala/io/indextables/spark/transaction/PartitionPredicateUtils.scala`**
   - Fix `resolveExpression` to handle numeric comparisons properly
   - Add numeric-aware comparison logic
   - Add support for additional operators (NOT, LIKE, etc.)

2. **`src/main/antlr4/io/indextables/spark/sql/parser/IndexTables4SparkSqlBase.g4`**
   - Consider replacing `.*?` with proper expression grammar
   - Or add explicit termination tokens for WHERE clauses

3. **`src/test/scala/io/indextables/spark/sql/DropPartitionsIntegrationTest.scala`**
   - Add tests for two-digit numeric partition values
   - Add tests for edge cases in comparison operators

4. **`src/test/scala/io/indextables/spark/sql/DropPartitionsParsingTest.scala`**
   - Add parsing tests for NOT, LIKE, RLIKE operators
   - Add tests for complex nested predicates

## Workarounds

Until fixed, users can work around these issues:

### For Numeric Comparisons

Use string literals with zero-padding:
```sql
-- Instead of:
WHERE month BETWEEN 1 AND 12

-- Use zero-padded strings:
WHERE month BETWEEN '01' AND '12'

-- Or use multiple OR conditions:
WHERE month IN ('1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12')
```

### For Date Ranges

Use string comparison with ISO format (which is lexicographically sortable):
```sql
-- This works because '2024-01-15' < '2024-02-01' lexicographically
WHERE date >= '2024-01-01' AND date <= '2024-12-31'
```

## Priority

**High** - This affects core functionality for partition management. Users may unknowingly corrupt their data by dropping/merging wrong partitions.

## Related

- The same predicate parsing is used in MERGE SPLITS and PREWARM CACHE commands
- Filter pushdown in read path (`FiltersToQueryConverter.scala`) may have similar issues but uses Spark's native filter handling

### PR DESCRIPTION
**Pull request by rbrenton**
## Summary

Fixes silent data correctness bug in partition predicate evaluation where numeric comparisons used lexicographic string ordering (`"10" < "2"` because `'1' < '2'`).

- **Root cause**: `PartitionPredicateUtils.resolveExpression` converted all non-string literals to `UTF8String`, causing Catalyst to use string comparison for what should be numeric comparisons
- **Secondary root cause**: Avro manifest partition pruner (`PartitionPruner`) computed partition bounds lexicographically, so `max("1","10","11","12","2",...,"9")` = `"9"` instead of `"12"`
- **7 affected code paths**: DROP PARTITIONS (destructive), MERGE SPLITS, PREWARM CACHE, DESCRIBE COMPONENT SIZES, BUILD COMPANION, prewarm partition filter, MERGE SPLITS countMergeGroups
- **Standard read path NOT affected**: `PartitionPruning.compareValues` already had type-aware comparison

## Changes

**Core fix** (`PartitionPredicateUtils.scala`):
- `buildPartitionSchema` now accepts optional full table schema from `MetadataAction.schemaString` to extract real partition column types
- `createRowFromPartitionValues` converts string partition values to their actual typed values (Int, Long, Float, Double, Boolean, Date, Timestamp, etc.)
- `resolveExpression` uses a two-pass approach: resolve attributes to typed BoundReferences, then coerce literals to match each column's actual data type (column-context-aware, not schema-global)
- `coerceLiteralToType` handles all partition-supported types: String, Integer, Long, Float, Double, Boolean, Short, Byte, Decimal, Date, Timestamp

**Schema caching** (`TransactionLog.scala`, `OptimizedTransactionLog.scala`, `TransactionLogFactory.scala`):
- Added `getSparkSchema()` and `getPartitionSchema()` with thread-safe double-checked locking cache
- Schema JSON is parsed once per TransactionLog lifetime, avoiding repeated 20-30ms parsing on wide tables
- All call sites simplified to `transactionLog.getPartitionSchema()`

**Avro manifest pruner fix** (`PartitionPruner.scala`, `AvroManifestWriter.scala`, `TombstoneDistributor.scala`):
- Added `numericAwareCompare` — tries Long, then Double parsing before falling back to lexicographic string comparison
- Applied to partition bounds computation (`computePartitionBounds`) and all bounds-check methods

**MergeSplitsCommand consolidation**: Removed ~96 lines of duplicate predicate evaluation code, replaced with delegation to `PartitionPredicateUtils`

**All 7 call sites updated**: DropPartitionsCommand, MergeSplitsCommand (2 paths), PrewarmCacheCommand (2 paths), DescribeComponentSizesCommand, IndexTables4SparkScan, SyncToExternalCommand

**Test fix**: `DescribeComponentSizesIntegrationTest` — updated non-existent path test to expect empty results instead of an exception (pre-existing failure on main)

## Test plan

- [x] 82 unit tests in `PartitionPredicateUtilsTest` — all pass (56 original + 26 new for Date, Timestamp, Boolean, Short, Byte types)
- [x] 16 integration tests in `DropPartitionsIntegrationTest` (including 3 IT-036 regression tests) — all pass
- [x] 16 parsing tests in `DropPartitionsParsingTest` — all pass
- [x] Full test suite (`make test`): 322/322 passed, 0 failures
- [x] Backward compatibility verified: string partition columns preserve lexicographic ordering
- [x] Type coercion tests with tightened assertions verify actual coerced literal types (not wildcards)

Closes #85

### CODE REVIEW

File: src/main/scala/io/indextables/spark/core/IndexTables4SparkScan.scala [hunk 1]
```diff
@@ -191,12 +191,9 @@ class IndexTables4SparkScan(
...
           try {
-            val metadata = transactionLog.getMetadata()
-            val partitionSchema = StructType(
-              metadata.partitionColumns.map(name =>
-                org.apache.spark.sql.types.StructField(name, org.apache.spark.sql.types.StringType, nullable = true)
-              )
-            )
+            val metadata        = transactionLog.getMetadata()
+            val fullSchema      = PartitionPredicateUtils.extractFullSchema(metadata)
+            val partitionSchema = PartitionPredicateUtils.buildPartitionSchema(metadata.partitionColumns, fullSchema)
```
Comment (schenksj) [inline src/main/scala/io/indextables/spark/core/IndexTables4SparkScan.scala]: This will require parsing the schema json multiple times thoughout the session lifecycle, which is quite expensive (20-30ms per call for very wide tables), and this is the hot path for transaction planning.  Can we please refactor this to assure that the full schema struct and partition schema are only calculated once in the lifetime of the transaciton log?  Perhaps it could be a lazy val in the transaction log itself, accessible via. transactionLog.getSparkSchema(), and transactionLog.getPartitionSchema()

File: src/main/scala/io/indextables/spark/transaction/PartitionPredicateUtils.scala [hunk 2]
```diff
@@ -237,16 +400,53 @@ object PartitionPredicateUtils {
...
+   * @param fullSchema
+   *   Optional full table schema (from MetadataAction.schemaString). If provided, partition columns will use their real
+   *   types. If None, falls back to StringType for all columns (legacy behavior).
    * @return
-   *   StructType with all columns as StringType
+   *   StructType with partition columns using their real types (or StringType if schema is unavailable)
    */
-  def buildPartitionSchema(partitionColumns: Seq[String]): StructType =
-    StructType(partitionColumns.map(name => StructField(name, StringType, nullable = true)))
+  def buildPartitionSchema(partitionColumns: Seq[String], fullSchema: Option[StructType] = None): StructType =
```
Comment (schenksj) [inline src/main/scala/io/indextables/spark/transaction/PartitionPredicateUtils.scala]: this feels like it belongs in the transaction log object rather than in a partition predicate utility class

**Review by schenksj — CHANGES_REQUESTED** [review]
Very nice!  One structural thing requested in the comments.  Thank you!

File: src/main/scala/io/indextables/spark/core/IndexTables4SparkScan.scala (hunk 1 above)
Comment (rbrenton) [reply src/main/scala/io/indextables/spark/core/IndexTables4SparkScan.scala]: Done — added `getSparkSchema()` and `getPartitionSchema()` to `TransactionLog` and `OptimizedTransactionLog` with thread-safe double-checked locking cache (`@volatile` + `synchronized`). Schema JSON is parsed once per TransactionLog lifetime. All 7 call sites now use `transactionLog.getPartitionSchema()`.

File: src/test/scala/io/indextables/spark/transaction/PartitionPredicateUtilsTest.scala [hunk 3]
```diff
@@ -451,4 +453,485 @@ class PartitionPredicateUtilsTest extends AnyFunSuite with Matchers {
       case other          => fail(s"Expected Not(And), got: $other")
     }
   }
+
+  // ==========================================================================
+  // IT-036: Type-Aware Partition Schema (buildPartitionSchema)
+  // ==========================================================================
+
+  test("buildPartitionSchema should use StringType when no full schema provided") {
```
Comment (schenksj) [inline src/test/scala/io/indextables/spark/transaction/PartitionPredicateUtilsTest.scala]: Please expand these tests to include partition's whose type are Date, Timestamp, TimestampNTZ, Boolean, Short, and Byte (I think these are all supported partition types.  I have some uncertainty about the type coersion method above, this should address it

File: src/main/scala/io/indextables/spark/transaction/PartitionPredicateUtils.scala [hunk 4]
```diff
@@ -124,53 +124,217 @@ object PartitionPredicateUtils {
...
+   * literals are converted to their string representation as UTF8String.
+   *
+   * @param literal
+   *   The literal to coerce
+   * @param targetType
+   *   The target data type (typically from a BoundReference)
+   * @return
+   *   A literal with a value compatible with the target type
+   */
+  private def coerceLiteralToType(literal: Literal, targetType: DataType): Literal =
```
Comment (schenksj) [inline src/main/scala/io/indextables/spark/transaction/PartitionPredicateUtils.scala]: I am uncertain if this is complete. was expecting to see Date, Timestamp, etc...  I left another comment in the tests section that might help guide the process.

**Review by schenksj — CHANGES_REQUESTED** [review]
Left a couple of related comments about data types.  thanks again!

File: src/main/scala/io/indextables/spark/transaction/PartitionPredicateUtils.scala (hunk 4 above)
Comment (rbrenton) [reply src/main/scala/io/indextables/spark/transaction/PartitionPredicateUtils.scala]: Done — added 6 missing type cases to `coerceLiteralToType`: BooleanType, ShortType, ByteType, DecimalType, DateType, and TimestampType. Now symmetric with `convertPartitionValue`. Each follows the existing pattern (UTF8String/toString extraction, try/catch with fallback). TimestampNTZ is Spark 3.4+ only, noted in tests.
